packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+
+      const indexParameterHasStringType = indexParameterTypeParts.some(
+        tsutils.isIntrinsicStringType,
+      );
+
+      // check remaining properties for a match
+      for (const key of remainingProperties.keys()) {
+        if (typeof key === 'number' && indexParameterHasNumberType) {
+          return true;
+        }
+
+        if (typeof key === 'string' && indexParameterHasStringType) {
+          return true;
+        }
+      }
+
+      // check dynamic properties for a match
+      for (const destructure of dynamicProperties) {
+        destructure.type ??= services.getTypeAtLocation(
+          destructure.property.key,
+        );
+
+        for (const type of tsutils.unionTypeParts(destructure.type)) {
+          if (checker.isTypeAssignableTo(type, indexParameterType)) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    /**
+     * @returns `true` if an index signature's value type matches an existing
+     * property's value type, `false` otherwise
+     */
+    function indexSignatureHasMatchingValue(
+      remainingProperties: RemainingProperties,
+      indexSignature: TSESTree.TSIndexSignature,
+    ): boolean {
+      // `{ used }: { [i: `_${string}`] }`
+      if (!indexSignature.typeAnnotation) {
+        return true;
+      }
+
+      const indexType = services.getTypeAtLocation(
+        indexSignature.typeAnnotation.typeAnnotation,
+      );
+
+      for (const destructure of remainingProperties.values()) {
+        destructure.type ??= services.getTypeAtLocation(
+          destructure.property.value,
+        );
+
+        if (isTypeMatchingIndexSignatureType(destructure.type, indexType)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    /**
+     * Compares an index-signature value-type with another type. If `isNoUncheckedIndexedAccess`
+     * is enabled, will filter out `undefined` from the index-signature's type.
+     */
+    function isTypeMatchingIndexSignatureType(
+      indexSignatureType: ts.Type,
+      comparedType: ts.Type,
+    ): boolean {
+      if (isNoUncheckedIndexedAccess) {
+        const indexSignatureParts = tsutils
+          .unionTypeParts(indexSignatureType)
+          .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+        const comparedTypeParts = tsutils.unionTypeParts(comparedType);
+
+        if (indexSignatureParts.length !== comparedTypeParts.length) {
+          return false;
+        }
+
+        const indexSignaturePartsSet = new Set(indexSignatureParts);
+
+        return comparedTypeParts.every(part =>
+          indexSignaturePartsSet.has(part),
+        );
+      }
+
+      return indexSignatureType === comparedType;
+    }
+
+    function reportOnMember(
+      member: TSESTree.Node,
+      data: { key: string; type: string },
+    ): void {
+      context.report({
+        node: member,
+        messageId: 'partialDestructuring',
+        data,
+        suggest: [
```

Comment:
> I included a suggestion to remove an unused property/element. Having this as an auto-fix seems inconvenient, as you sometimes write the type before you destructure over it.

[Feature] I'd personally prefer an autofix. There already exist a lot of rules that can inconveniently change WIP code. For example, [`@typescript-eslint/consistent-indexed-object-style`](https://typescript-eslint.io/rules/consistent-indexed-object-style) can autofix an index signature to a `Record` if you run it before adding another property to the index signature. Rules aren't intended to be run on WIP code.

Also: when I'm newly enabling this rule on a codebase with many violations, I would sure love an autofixer!

Reply:
I'm convinced on this, thanks 👍